subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -203,26 +206,35 @@ public void beforeComplete(GradleInternal gradle) {
             return locationsWrittenByCurrentBuild;
         }
 
+        WatchableFileSystemDetector createWatchableFileSystemDetector() {
+            return DefaultWatchableFileSystemDetector.create();
+        }
+
+        WatchableHierarchies createWatchableHierarchies(GlobalCacheLocations globalCacheLocations, WatchableFileSystemDetector watchableFileSystemDetector) {
```

Comment:
I don't think this can be service, since it contains state. And we currently wouldn't update it if watching is disabled.

Before your change, the `WatchableHierarchies` would be created every time we start watching and destroyed every time we stopped watching. Now they are kept around, even if we stop watching. We need to either make lifecycle calls on the watchable hierarchies or not keep it as a service.

Reply:
Good point, fixed. These should be build-scoped services then, though, no?

Reply:
No. They live between builds if file system watching is started. They are "FileWatcherRegistry" scoped services. As long as we keep the registry around, we also need to keep the `WatchableHierarchies` around.

Reply:
Right. I guess we should capture this scope instead of `new`-ing up services as needed.

Reply:
Yeah, let's look at this together next week.